File: gui/mainwindow.rb

#require 'gui/customer'
#require 'gui/invoice'
#require 'data/storage'
require 'gui/commons'

class MainWindow < Qt::MainWindow

	def initialize()
		super()

		if Qt::Icon::theme_name.eql? "hicolor"
			Qt::Icon::set_theme_name('Tango')
		end

		@i2p = nil

		init_actions
		init_gui
		init_connections
		revalidate
	end

	protected
	def init_actions
		@action_quit = Qt::Action.new(Qt::Icon::fromTheme('exit'), 'Quit', self)
		@action_quit.connect(SIGNAL :triggered) { $qApp.quit }
	end

	def init_gui
		self.set_minimum_size(600,300)
		
		menu_app = self.menu_bar.add_menu('Images2PDF'.qtr)
		menu_app.add_action(@action_quit)

		widget_main = Qt::Widget.new(self)
		layout_main = Qt::FormLayout.new(widget_main)

		@folder = FileInput.new FileInput::MODE_DIR
		@folder.dialog_caption = 'Images directory'.qtr
		@file = FileInput.new FileInput::MODE_FSAVE
		@file.dialog_caption = 'Save as PDF'.qtr

		@folder.value = '/tmp/i2p/small'
		@file.value = '/tmp/res.pdf'

		@process_button = Qt::PushButton.new('Merge'.qtr)

		@status = Qt::Label.new

		@progress = Qt::ProgressBar.new
		@progress.minimum = 0
		@progress.visible = false

		layout_main.addRow('Folder with images:', @folder)
		layout_main.addRow('Target pdf file:', @file)
		layout_main.addRow(@process_button)
		layout_main.addRow(@status)
		layout_main.addRow(@progress)

		self.setCentralWidget(widget_main)
	end

	def init_connections
		@folder.connect(SIGNAL :value_changed) { self.revalidate }
		@file.connect(SIGNAL :value_changed) { self.revalidate }
		@process_button.connect(SIGNAL :clicked) { self.process_start }
	end

	def revalidate
		if @folder.readable_dir? and @file.writable_file?
			@status.text = 'You can press Merge button now.'.qtr
			@process_button.enabled = true
		else
			@status.text = 'Select path to dir with images and where to save pdf.'.qtr
			@process_button.enabled = false
		end
	end

	def process_start
		@status.text = 'Preparing'.qtr

		@progress.value = 0
		@progress.maximum = 1
		@progress.visible = true

		@folder.enabled = false
		@file.enabled = false
		@process_button.enabled = false

		@i2p = Images2PDF.new(@folder.value, @file.value)
		@i2p.callback = Proc.new { |tot, done, last| self.process_step(tot, done, last) }
		@i2p.callback_finishing = Proc.new { self.process_rendering }
		@i2p.callback_finished = Proc.new { self.process_end }
		@i2p.start
	end

	def process_step(total, finished, current_file)
		@progress.maximum = total
		@progress.value = finished
		@status.text = "Added #{current_file}"
	end

	def process_rendering
		@progress.visible = false
		@status.text = "Rendering pdf please wait ..."
	end

	def process_end
		@i2p = nil

		@progress.visible = false

		@folder.enabled = true
		@file.enabled = true
		@process_button.enabled = false

		@status.text = 'All done. Hope you are satisfied with results.'
	end

end

File: mainwindow.cpp
#include "mainwindow.h"
#include <QFormLayout>
#include <QDialog>
#include <QApplication>
#include <QIcon>
#include <QMenuBar>
#include <QMenu>

MainWindow::MainWindow(QWidget *parent) :
    QMainWindow(parent),
    _convertor(NULL)
{
    // workaround because themed icons dont work in fluxbox as they should
    if(QIcon::themeName().isEmpty() || QIcon::themeName() == "hicolor") {
        QIcon::setThemeName("Tango");
    }

    _init_actions();
    _init_gui();
    _init_connections();

    revalidate();
}

MainWindow::~MainWindow()
{
}

void MainWindow::_init_actions()
{
    _quit = new QAction(QIcon::fromTheme("exit"), trUtf8("Quit"), this);
}

void MainWindow::_init_gui()
{
    setMinimumSize(600, 200);

    QMenu *menu = this->menuBar()->addMenu(trUtf8("Images2PDF"));
    menu->addAction(_quit);

    QWidget *cw = new QWidget();
    setCentralWidget(cw);
    QFormLayout *layout = new QFormLayout(cw);

    _folder = new FileInput(FileInput::DirectoryMode);
    _folder->dialog_caption = trUtf8("Images directory");
    _file = new FileInput(FileInput::FileSaveMode);
    _file->dialog_caption = trUtf8("Save as PDF");

    _process = new QPushButton(trUtf8("Merge"));
    _abort = new QPushButton(trUtf8("Abort"));
    _status = new QLabel();
    _progress = new QProgressBar();

    _abort->setVisible(false);

    _progress->setMinimum(0);
    _progress->setVisible(false);

    layout->addRow(trUtf8("Folder with images"), _folder);
    layout->addRow(trUtf8("Target pdf file:"), _file);
    layout->addRow(_process);
    layout->addRow(_abort);
    layout->addRow(_status);
    layout->addRow(_progress);
}

void MainWindow::_init_connections()
{
    connect(_folder, SIGNAL(valueChanged()), this, SLOT(revalidate()));
    connect(_file, SIGNAL(valueChanged()), this, SLOT(revalidate()));

    connect(_quit, SIGNAL(triggered()), qApp, SLOT(quit()));
    connect(_process, SIGNAL(clicked()), this, SLOT(process()));
    connect(_abort, SIGNAL(clicked()), this, SLOT(processAbort()));
}

void MainWindow::revalidate()
{
    qDebug() << "revalidating";
    if(_folder->isReadableDir() && _file->isWritableFile()) {
        _status->setText(trUtf8("You can press Merge button now."));
        _process->setEnabled(true);
    } else {
        _status->setText(trUtf8("Select path to dir with images and where to save pdf."));
        _process->setEnabled(false);
    }
}

void MainWindow::process()
{
    if(_convertor) {
        return;
    }
    _status->setText(trUtf8("Preparing ..."));

    _folder->setEnabled(false);
    _file->setEnabled(false);
    _process->setEnabled(false);

    _convertor = new Images2PDF(_folder->value(), _file->value());

    _status->setText(trUtf8("Preparing to process %1 images").arg(_convertor->imageCount()));
    _progress->setValue(0);
    _progress->setMaximum(_convertor->imageCount());
    _progress->setVisible(true);

    _abort->setVisible(true);

    connect(_convertor, SIGNAL(processing(QString)), this, SLOT(processProcessing(QString)));
    connect(_convertor, SIGNAL(progress(int)), this, SLOT(processProgress(int)));
    connect(_convertor, SIGNAL(finished(bool)), this, SLOT(processFinished(bool)));

    _convertor->start();
}

void MainWindow::processFinished(bool success)
{
    _folder->setEnabled(true);
    _file->setEnabled(true);
    _process->setEnabled(true);

    _abort->setVisible(false);;
    _progress->setVisible(false);

    if(success) {
        _status->setText(trUtf8("All done, enjoy your pdf ;)"));
    } else if(_convertor) {
        _status->setText(_convertor->error());
    }
    if(_convertor) {
        delete _convertor;
        _convertor = NULL;
    }
}

void MainWindow::processProcessing(QString fileName)
{
    _status->setText(trUtf8("Processing %1").arg(fileName));
}

void MainWindow::processProgress(int done)
{
    _progress->setValue(done);
}

void MainWindow::processAbort()
{
    if(_convertor) {
        _convertor->abort();
    }
}
